### ISSUE TITLE
Clang -pedantic: non-ISO practices may not be diagnosed

### ISSUE THREAD

**Issue by pmor13** [llvm/llvm-project#143486]
I do believe that per ISO C11 this code :
```
int x __attribute__ ((aligned (16))) = 0;
```
leads to syntax error.

A conforming implementation shall produce at least one diagnostic message.

Clang with options `-std=c11 -Wall -Wextra -pedantic` produces nothing.

@AaronBallman Any comments?

**Comment by pmor13** [comment]
Expected diagnostics:
```
warning: ISO C does not support '__attribute__((xxx))' attributes [-Wpedantic]
```

**Comment by frederick-vs-ja** [comment]
Looks weird. IIUC implementations expect that `__attribute__` (together with `__declspec`) is allowed to be a silent extension. Perhaps the C standard needs to be fixed.

The C++ standard renders using of such an extension in user-code ill-formed, no diagnostic required. This also sounds somehow weird, but at least allows implementations to be silent on `__attribute__`.

**Comment by AaronBallman** [comment]
There are cases where we intentionally do not warn on use of extensions, and attribute syntax is one of them (we don't warn on `__declspec` or `__attribute__` as extensions). The trouble is: there are a *lot* of attributes and they're often hidden behind macros. The purpose of `-pedantic` is to tell the user when they're using an extension because that's inherently not portable. But macros are used to make attributes portable -- you have fallback expansions for other compilers with different syntax or a different set of attributes. So many of the diagnostics we'd emit would not be actionable for the user and suppressing those diagnostics on a case-by-case basis would be really frustrating. So we'd be adding a new diagnostic group with the expectation that users would be disabling that warning most of the time. It's not a good value proposition in a lot of ways.

FWIW, we will *never* emit a pedantic diagnostic for every extension, regardless of what standard conformance requires. Consider the `__extension__` extension: the whole point to that extension is to silence extension warnings. Emitting an extension warning for that would defeat the entire purpose of the feature, even though we technically are required to issue a diagnostic for its use because of the invalid syntax (according to the standard).

**Comment by AaronBallman** [comment]
> The C++ standard renders using of such an extension in user-code ill-formed, no diagnostic required. This also sounds somehow weird, but at least allows implementations to be silent on `__attribute__`.

Not according to my reading of the standard:
https://eel.is/c++draft/intro.compliance#general-1
https://eel.is/c++draft/intro.compliance#general-2.3.1
https://eel.is/c++draft/intro.compliance#general-10

It's invalid syntax (not explicitly specified as undefined behavior anywhere I can find), so it's a diagnosable rule, so in theory we're supposed to emit a diagnostic. What's more, "provided they do not alter the behavior of any well-formed program" in p10 is something we happily ignore without emitting a diagnostic, too. We have a number of attributes which impact program correctness on code which would be well-formed without the attribute (consider things like alignment attributes via non-standard means, as an example).

**Comment by frederick-vs-ja** [comment]
> > The C++ standard renders using of such an extension in user-code ill-formed, no diagnostic required. This also sounds somehow weird, but at least allows implementations to be silent on `__attribute__`.
> 
> Not according to my reading of the standard: https://eel.is/c++draft/intro.compliance#general-1 https://eel.is/c++draft/intro.compliance#general-2.3.1 https://eel.is/c++draft/intro.compliance#general-10
> 
> It's invalid syntax (not explicitly specified as undefined behavior anywhere I can find), so it's a diagnosable rule, so in theory we're supposed to emit a diagnostic. What's more, "provided they do not alter the behavior of any well-formed program" in p10 is something we happily ignore without emitting a diagnostic, too. We have a number of attributes which impact program correctness on code which would be well-formed without the attribute (consider things like alignment attributes via non-standard means, as an example).

Ah, however, it seems that IFNDR-ness has higher priority - p2.2 applies relaxation for IFNDR first, and then p2.3 places diagnostic-requiring cases in "otherwise"  (also mentioned in https://github.com/cplusplus/CWG/issues/357#issuecomment-1630712012). And [[lex.name] p3](https://eel.is/c++draft/lex.name#3) indicates that use of `__attribute__`, `__declspec`, or `__extension__` is IFNDR...

**Comment by AaronBallman** [comment]
Ah! I missed the NDR part in lex.name, thanks!

**Comment by pmor13** [comment]
@AaronBallman 
> a lot of attributes and they're often hidden behind macros.

I didn't know about that. Can you please provide an example of such macro(s)?

> a new diagnostic group with the expectation that users would be disabling that warning most of the time.

What about adding `-more-pedantic` (or similar name) option (diagnose usage of all extensions)?

**Comment by AaronBallman** [comment]
> [@AaronBallman](https://github.com/AaronBallman)
> 
> > a lot of attributes and they're often hidden behind macros.
> 
> I didn't know about that. Can you please provide an example of such macro(s)?

Sure! There are quite a few examples in here: https://github.com/llvm/llvm-project/blob/e74d834cb155a894fa0f9dbd1483b7fef53a79ae/llvm/include/llvm/Support/Compiler.h#L117

> > a new diagnostic group with the expectation that users would be disabling that warning most of the time.
> 
> What about adding `-more-pedantic` (or similar name) option (diagnose usage of all extensions)?

Can you explain the use case for the diagnostic a bit more and what you would expect the diagnostic to cover? Would you expect it to diagnose use of `__extension__`? Or `__INT_MAX__`? Or `__builtin_*`? What about non-standard attributes or pragmas? Do you expect it to diagnose IFNDR constructs in C++?

**Comment by pmor13** [comment]
@AaronBallman Thanks for the example.

> Can you explain the use case for the diagnostic a bit more and what you would expect the diagnostic to cover?

Perhaps the most significant use case that quickly comes to mind is checking whether a program is maximally portable (among conforming implementations). Such program 1) uses only those features of the language and library specified in the Standard, and, hence, 2) doesn't use any nonportable features and extensions.

I expect the diagnostic to cover usage of all nonportable features and extensions _in the user's code_ (i.e. not, for example, in the standard headers provided by an implementation).

> Would you expect it to diagnose use of __extension__? Or __INT_MAX__? Or __builtin_*?

Only of these are used in the user's code. (Their usage, for example, in the standard headers provided by an implementation is OK, I suppose.)

> What about non-standard attributes or pragmas?

Do you mean `[[xxx]]` under non-standard attributes? I suppose that any "non-standard <anything>" needs (in this "the strictest standard conformance diagnostic mode") to be diagnosed unless the standard explicitly gives a permission to ignore (see below) such "non-standard <anything>". BTW, does C23 and / or C++23 require non-standard attributes to be diagnosed?

Non-standard pragmas need not be diagnosed: per C11 6.10.6 Pragma directive:
> Any such pragma that is not recognized by the implementation is ignored.

I suppose that the benefit (?) of `#pragma` over extension (usage of which violates any syntax rule or constraint) is that the recognized pragma need not be diagnosed, while an extension (usage of which violates any syntax rule or constraint) is required to be diagnosed.

> Do you expect it to diagnose IFNDR constructs in C++?

I'm not very familiar with C++. I suppose that yes, all IFNDR constructs need (in this "the strictest standard conformance diagnostic mode") to be diagnosed.

P.S. The `-more-pedantic` may be named `-very-pedantic`. If adding new option is considered tiresome and / or boring, then the proposed "the strictest standard conformance diagnostic mode" checks may be enabled as a part of `-Weverything` option. However, `-Weverything` it's not recommended to be used by the end user: [1](https://softwareengineering.stackexchange.com/a/124574/366642), [2](https://quuxplusone.github.io/blog/2018/12/06/dont-use-weverything/).

**Comment by AaronBallman** [comment]
> [@AaronBallman](https://github.com/AaronBallman) Thanks for the example.
> 
> > Can you explain the use case for the diagnostic a bit more and what you would expect the diagnostic to cover?
> 
> Perhaps the most significant use case that quickly comes to mind is checking whether a program is maximally portable (among conforming implementations). Such program 1) uses only those features of the language and library specified in the Standard, and, hence, 2) doesn't use any nonportable features and extensions.
>
> I expect the diagnostic to cover usage of all nonportable features and extensions _in the user's code_ (i.e. not, for example, in the standard headers provided by an implementation).

I think we'll have to read into this a bit to get to something even close to useful in practice though. You're already carving out standard headers, but what about other system headers? `#include <Windows.h>` is not portable, for example. But even beyond that, "maximally portable" is not a goal I think most users want when they realize what that actually means. For example, in C, any header file you include that's more than 8 characters before the period, has more than one character after the period, or has no period is not portable. Use of `long` is not portable because different targets make different choices as to its width. On Windows, the entrypoint for a GUI application is `WinMain` rather than `main` and that's not portable. And so on.

> > Would you expect it to diagnose use of **extension**? Or **INT_MAX**? Or __builtin_*?
> 
> Only of these are used in the user's code. (Their usage, for example, in the standard headers provided by an implementation is OK, I suppose.)

System headers use these sort of constructs all the time though, and those aren't provided by the implementation.

> > What about non-standard attributes or pragmas?
> 
> Do you mean `[[xxx]]` under non-standard attributes? I suppose that any "non-standard " needs (in this "the strictest standard conformance diagnostic mode") to be diagnosed unless the standard explicitly gives a permission to ignore (see below) such "non-standard ". BTW, does C23 and / or C++23 require non-standard attributes to be diagnosed?

The request was for maximal portability though? The standards don't require non-standard attributes to be diagnosed, but their use is still non-portable.

> Non-standard pragmas need not be diagnosed: per C11 6.10.6 Pragma directive:
> 
> > Any such pragma that is not recognized by the implementation is ignored.

But that's still not a maximally portable construct, same as attributes.

> I suppose that the benefit (?) of `#pragma` over extension (usage of which violates any syntax rule or constraint) is that the recognized pragma need not be diagnosed, while an extension (usage of which violates any syntax rule or constraint) is required to be diagnosed.
> 
> > Do you expect it to diagnose IFNDR constructs in C++?
> 
> I'm not very familiar with C++. I suppose that yes, all IFNDR constructs need (in this "the strictest standard conformance diagnostic mode") to be diagnosed.
> 
> P.S. The `-more-pedantic` may be named `-very-pedantic`. If adding new option is considered tiresome and / or boring, then the proposed "the strictest standard conformance diagnostic mode" checks may be enabled as a part of `-Weverything` option. However, `-Weverything` it's not recommended to be used by the end user: [1](https://softwareengineering.stackexchange.com/a/124574/366642), [2](https://quuxplusone.github.io/blog/2018/12/06/dont-use-weverything/).

I think your request is valid insomuch as there are extensions we don't diagnose that we should. However, I really don't think we'd want to go so far as to emit a diagnostic for every extension. That introduces compile time overhead for very little value in practice. I think a better approach is to explore which extensions we're not issuing diagnostics for and take them on a case by case basis as to whether they should be diagnosed. For example, I'm strongly opposed to any extension diagnostic for `__extension__`; the whole point to that construct is to silence extension warnings. But anything which is IFNDR that we can reasonably diagnose should be diagnosed, that code is ill-formed after all. If we're failing to warn on some backported feature, that's a good candidate for us to diagnose, etc.

At the end of the day, I think the only realistic way to actually learn about maximal portability of any piece of code is to try compiling that code with the toolchain you want to be compatible with. We're never going to get it "perfect" from our end, there are too many differences between targets, architectures, compilers, etc. If we're diagnosing everything that could possibly be different, you will be swamped with diagnostics and very little ability to take action on them.

**Comment by pmor13** [comment]
@AaronBallman 

1. About "diagnose IFNDR constructs in C++". I provided wrong answer, sorry. Below is the correct one.

> Do you expect it to diagnose IFNDR constructs in C++?

No, because of NDR. The rationale for this answer is the following: "diagnose what is required, do not diagnose what is not required". However, for the sake of high quality-of-implementation such IFNDR constructs in C++ may be diagnosed.

2. What if we narrow the goal down to only "1) uses only those features of the language and library specified in the Standard"? In other words: diagnose when a program does not "use only those features of the language and library specified in the Standard".

3. Re: "If we're diagnosing everything...": that "diagnosing everything" is a special mode (disabled by default).
4. What if we add support for `#pragma "check that program uses only those features of the language and library specified in the Standard" ON / OFF`? Example:
```
#include <stdio.h>
#pragma clang_check_strict_iso23_conformance ON
<some code>
#pragma clang_check_strict_iso23_conformance OFF
```
